language-plutus-core/stdlib/Language/PlutusCore/StdLib/Data/Function.hs

```diff
@@ -92,3 +95,155 @@ getBuiltinFix = do
         $ [ Apply () (unrollFunAB unroll2) $ Var () s
           , Var () x
           ]
+
+-- | A type that looks like a transformation.
+--
+-- > trans F G Q : F Q -> G Q
+trans :: Type TyName () -> Type TyName () -> Type TyName () -> Quote (Type TyName ())
+trans f g q = pure $ TyFun () (TyApp () f q) (TyApp () g q)
+
+-- | A type that looks like a natural transformation, sometimes written 'F ~> G'.
+--
+-- > natTrans F G : forall Q :: * . F Q -> G Q
+natTrans :: Type TyName () -> Type TyName () -> Quote (Type TyName ())
+natTrans f g = freshTyName () "Q" >>= \q -> TyForall () q (Type ()) <$> trans f g (TyVar () q)
+
+-- | A type that looks like a natural transformation to Id.
+--
+-- > natTransId F : forall Q :: * . F Q -> Q
+natTransId :: Type TyName () -> Quote (Type TyName ())
+natTransId f = do
+    q <- freshTyName () "Q"
+    pure $ TyForall () q (Type ()) (TyFun () (TyApp () f (TyVar () q)) (TyVar () q))
+
+-- | The 'fixBy' combinator.
+--
+-- fixBy :
+--     forall (F :: * -> *) .
+--     ((F ~> Id) -> (F ~> Id)) ->
+--     ((F ~> F) -> (F ~> Id))
+getBuiltinFixBy :: Quote (Term TyName Name ())
+getBuiltinFixBy = do
+    f <- freshTyName () "F"
+
+    -- by : (F ~> Id) -> (F ~> Id)
+    by <- freshName () "by"
+    byTy <- do
+        nt1 <- natTransId (TyVar () f)
+        nt2 <- natTransId (TyVar () f)
+        pure $ TyFun () nt1 nt2
+
+    -- instantiatedFix = fix {F ~> F} {F ~> Id}
+    instantiatedFix <- do
+        fix <- getBuiltinFix
+        nt1 <- natTrans (TyVar () f) (TyVar () f)
+        nt2 <- natTransId (TyVar () f)
+        pure $ TyInst () (TyInst () fix nt1) nt2
+
+    -- rec : (F ~> F) -> (F ~> Id)
+    recc <- freshName () "rec"
+    reccTy <- do
+      nt <- natTrans (TyVar () f) (TyVar () f)
+      nt2 <- natTransId (TyVar () f)
+      pure $ TyFun () nt nt2
+
+    -- h : F ~> F
+    h <- freshName () "h"
+    hty <- natTrans (TyVar () f) (TyVar () f)
+
+    -- R :: *
+    -- fr : F R
+    r <- freshTyName () "R"
+    fr <- freshName () "fr"
+    let frTy = TyApp () (TyVar () f) (TyVar () r)
+
+    -- Q :: *
+    -- fq : F Q
+    q <- freshTyName () "Q"
+    fq <- freshName () "fq"
+    let fqTy = TyApp () (TyVar () f) (TyVar () q)
+
+    -- inner = (/\ Q :: * -> \ q : F Q -> rec h {Q} (h {Q} q))
```

Comment:
Formatting nitpicking again: `/\ (Q :: *) -> ...` is more standard and reads better I think.